src/fsharp/MethodOverrides.fs

```diff
@@ -420,7 +459,116 @@ module DispatchSlotChecking =
+            ([ minfo1 ], sortedPossibleOverriderMethodsTail)
+            ||> List.fold (fun minfos minfo2 ->
+                let isDiamondHierarchy =
+                    not (TypeFeasiblySubsumesType 0 g amap m minfo1.ApparentEnclosingType CanCoerce minfo2.ApparentEnclosingType) && 
+                    not (TypeFeasiblySubsumesType 0 g amap m minfo2.ApparentEnclosingType CanCoerce minfo1.ApparentEnclosingType)
+
+                if isDiamondHierarchy then
+                    minfo2 :: minfos
+                else
+                    minfos
+            )
+
+    /// Get a collection of slots for the given interface type.
+    let GetInterfaceDispatchSlots (infoReader: InfoReader) ad m availImpliedInterfaces reqdTyInfos tys =
+        let g = infoReader.g
+        let amap = infoReader.amap
+
+        [ for ty in tys do
+            if isInterfaceTy g ty then
+                // Check if the interface has an inherited implementation
+                // If so, you do not have to implement all the methods - each
+                // specific method is "optionally" implemented.
+                let isOptional = ListSet.contains (typeEquiv g) ty availImpliedInterfaces
+                for minfo in GetImmediateIntrinsicMethInfosOfType (None, ad) g amap m ty do
+                  if minfo.IsNewSlot then
+                      if isOptional || not minfo.IsILMethod then
+                          yield RequiredSlot(minfo, GetDefaultDispatchSlotFlags minfo ||| (if isOptional then RequiredSlotFlags.Optional else RequiredSlotFlags.None))
+                      else
+                          let dispatchFlags =
+                              match GetTopMostHierarchicalInterfaceOverriderMethods infoReader m reqdTyInfos minfo with
+                              | [] -> GetDefaultDispatchSlotFlags minfo
+                              | [ minfo2 ] -> GetDefaultDispatchSlotFlags minfo2
+                              | _ -> RequiredSlotFlags.PossiblyNoMostSpecificImplementation
+
+                          yield RequiredSlot (minfo, dispatchFlags) ]
+
+    /// Get a collection of slots for the given class type.
+    let GetClassDispatchSlots (infoReader: InfoReader) ad m reqdTy =
+        let g = infoReader.g
+
+        [ if not (isInterfaceTy g reqdTy) then
+            // In the normal case, the requirements for a class are precisely all the abstract slots up the whole hierarchy.
+            // So here we get and yield all of those.
+            for minfo in reqdTy |> GetIntrinsicMethInfosOfType infoReader None ad AllowMultiIntfInstantiations.Yes IgnoreOverrides m do
+                if minfo.IsDispatchSlot then
+                    yield RequiredSlot(minfo, GetDefaultDispatchSlotFlags minfo) ]
+
+    /// Get a collection of slots for the given type and implied types.
+    let GetDispatchSlots (infoReader: InfoReader) ad m availImpliedInterfaces reqdTyInfos reqdTy impliedTys =
+        let g = infoReader.g
+
+        [ if isInterfaceTy g reqdTy then 
```

Comment:
I think this is the same as
```fsharp
        if isInterfaceTy g reqdTy then 
            GetInterfaceDispatchSlots infoReader ad m availImpliedInterfaces reqdTyInfos impliedTys
          else                  
            GetClassDispatchSlots infoReader ad m reqdTy
```
without the list expression

Reply:
Yea, we don't need one here now. :)